text/0025-confidential-token-ids.md

```diff
@@ -0,0 +1,220 @@
+    - Note also, the blinding factor base point does not need to be unique per `token_id`, so we use the same blinding factor base point for all `token_ids`.
+
+### Fee Outputs
+
+Currently, the enclave aggregates the fees for multiple transactions in a block, in order to mint a single fee output for the block and conserve space. To support multiple confidential `token_ids`, when minting fee outputs, the enclave must create multiple fee outputs if multiple `token_ids` were used in the block. A consideration here is that we would like for an observer not to discern whether a block contains transactions of multiple asset types by statistically analyzing the number of outputs to derive information about the number of fee outputs. 
+
+A simple proposal is that each block now contains a number of fee outputs scaling linearly with the total number of supported `token_ids`. This way we can continue to aggregate fees, but not reveal how many token types were included in the block. For blocks with fewer transactions than `token_ids`, the number of fee outputs is `min(num_token_ids, num_transactions_in_block)`
+
+Note that the untrusted portion of the node software knows how many transactions are in each block, and this is not considered statistically relevant information with respect to confidentiality. In other words, the current fee aggregation behavior is not meant as a confidentiality measure, purely as a space-saving measure.
+
+### Fee Priority and Transaction Sorting
+
+Currently the [WellFormedTxContext](https://github.com/mobilecoinfoundation/mobilecoin/blob/a092039a4a5cc7f5e3e16eeadd0b2bc3a12667ae/consensus/enclave/api/src/lib.rs#L50) contains the fee, which must be used by untrusted to [sort transactions](https://github.com/mobilecoinfoundation/mobilecoin/blob/a092039a4a5cc7f5e3e16eeadd0b2bc3a12667ae/consensus/enclave/api/src/lib.rs#L131) when [combining transactions](https://github.com/mobilecoinfoundation/mobilecoin/blob/a092039a4a5cc7f5e3e16eeadd0b2bc3a12667ae/consensus/service/src/tx_manager/untrusted_interfaces.rs#L40) for consideration by the enclave, prioritizing higher fees for inclusion in the block during periods of network congestion. In order to keep the `token_id` confidential, we must provide a mechanism for untrusted to accurately sort transactions before asking the enclave to do work on the transactions, without revealing the `token_id`, and at the same time, allowing for different asset types to express differing minimum fee requirements.
+
+To derive a sorting order without revealing the transaction type, the `fee` in the `WellFormedTxContext` derives a `priority`, which is calculated based on some ratio between the minimum fees configured by the node operator on startup. For example, the network operators may have configured the following minimum fees:
```

Comment:
This may allow node operators to set up a token-id oracle.

Reply:
We're trying to avoid that by normalizing based on minimum fee, usually the users just always use the minimum fee, and then priority would reveal no information.

It is true that we would not have this problem if MOB were the only thing that could be used for fees.

Reply:
A node operator can set a tx's priority to an arbitrary value by configuring the minimum fees on startup. This lets them guess-and-check the sorting to find the tx's token id.

Reply:
> A node operator can set a tx's priority to an arbitrary value by configuring the minimum fees on startup

That's true but their node can't consense with other nodes unless they have the same configuration as them, because the config is hashed and then appended to the responder id used for node-to-node attestation. This means the the nodes won't be able to pair to eachother unless they have the same config.

So in real life, the attacker has to hard fork the network, and control all the nodes that the user is submitting to, in order to conduct this guess-and-check attack.

Still, I agree it would be better if this weren't possible at all.

Reply:
I see, then I think this part of the document should be clarified, since the way it's written it sounds like nodes can freely define their own startup values.